core/src/main/java/org/elasticsearch/node/Node.java

```diff
@@ -282,28 +289,73 @@ public Node start() {
         injector.getInstance(MonitorService.class).start();
         injector.getInstance(RestController.class).start();
 
+        assert injector.getInstance(ClusterService.class) instanceof InternalClusterService :
+                "node cluster service implementation must inherit from InternalClusterService";
```

Comment:
for my education, does it make sense to have the interface if we require the implementation?